### ISSUE TITLE
3DFX in 'Lands of Lore - Guardians of Destiny (1997)'

### ISSUE THREAD

**Issue by Python-Exoproject** [dosbox-staging/dosbox-staging#3481]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.81

### What Operating System are you using?

Windows 11

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Recently it was discovered that the game crashes with a memory allocation error (as per below image) when you trigger a speech by the character Baccata. 

This crash was traced to the presence of the 'glide2x.ovl' file but I have not been able to find a workaround that will enable the game to be successfully played in 3DFX mode. 

![image](https://github.com/dosbox-staging/dosbox-staging/assets/67406191/5c5685cb-0e8f-49d3-8815-887c0911fc1e)

### Steps to reproduce the behaviour.

For testing here is a save provided by a member of the exodos discord: [CRASH_T.zip](https://github.com/dosbox-staging/dosbox-staging/files/14526653/CRASH_T.zip). 

To trigger the crash "Load save, go around corner to cantina. Click on door after cutscene plays."

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
I guess the question is has the game ever not crashed on real hardware with Voodoo in the same scenario? That would be very useful to find out before we start chasing our tails. E.g., does DOSBox X or PCem not crash? That at least would give us some data points.

**Comment by johnnovak** [comment]
> I can tell you that ECE and X also crash the same way, not sure about PCem.

That's useful info, cheers.

> I believe the 3DFX patch is an official one so theoretically went through QA before it was released, but it wouldnt be unheard of for something to be released broken. Hopefully someone with hardware setup will be able to test.

Yeah, especially if talking to that character is not essential for completing the game. Probably this will be on hold until someone can confirm the behaviour on hardware.

**Comment by Grounded0** [comment]
Is the game patched to the latest version 1.30?:

https://www.patches-scrolls.de/patch/2407/7/

**Comment by Python-Exoproject** [comment]
The eXoDOS version is patched, yes

**Comment by boohyaka** [comment]
Looks like issue is somewhat known:

https://www.gog.com/forum/lands_of_lore_series/lands_of_lore_2_glide_and_general_configuration_guide_plus_troubleshooting

> HELP! I crash to the desktop when chatting with Baccata in the Huline Jungle Cantina
> When/If you get the Cantina crash to desktop bug, edit the dosboxLOL2_single.conf
> 
> #For normal gameplay
> lolg.exe -CD ..
> 
> #Cantina Crash, Remove '#' from following line and add '#' it to the preceeding line. Switch back after bypassing bug
> #lolg.exe -LOWMEM -CD ..
> TO
> 
> #For normal gameplay
> #lolg.exe -CD ..
> 
> #Cantina Crash, Remove '#' from following line and add '#' it to the preceeding line. Switch back after bypassing bug
> lolg.exe -LOWMEM -CD ..
> This should enable you to get past that scene. After saving past that point. Quit, reverse the changes to dosboxLOL2_single.conf.
> 
> You'll also need to reset texture cache to large in the game's video options, and may need to change graphics options back to high if you disable hardware acceleration 

I'm trying to test it on my real hardware but the game is being an ass right now. Will report when I'm able to test.

**Comment by johnnovak** [comment]
> I'm trying to test it on my real hardware but the game is being an ass right now. Will report when I'm able to test.

Excellent @boohyaka, great find. Curious of your test tesults on real hardware.

**Comment by boohyaka** [comment]
still unable to run the game, gah. will try again over the weekend. in the meantime after some reading it is indeed a known issue, but always in emulation environments, haven't been able to find people confirming it runs on real hardware.

Apparently the LOWMEM parameter just reduces graphic quality. And it seems people running DOSBox in windowed mode also don't have the issue. Maybe something you're able to confirm @Python-Exoproject?

**Comment by Grounded0** [comment]
If the game doesn't crash with `LOWMEM` specified have you tried to run the game with 32 MB of RAM? DOS extenders usually don't support virtual memory so the memory you have is the memory that can be allocated.

Fallout will crash eventually unless it has 32 MB on DOS while on Windows it runs on less as it can swap into virtual memory.

**Comment by Grounded0** [comment]
The moment of crash recorded:

https://youtu.be/CkqWPbIAN7g?si=CmEth6Dk6aZrEWSn&t=1063

**Comment by Python-Exoproject** [comment]
@boohyaka I only test in windowed, so can confirm window mode makes no difference. Like you ive seen the odd reference to the issue online but never with real hardware, always with an emulated environment. 

Another possibly relevant bit of behaviour that seems unique to emulated environments - the presence of the glide2x.ovl file causes the same issue just by being present even if you have `hardware acceleration=off` in the games settings.

Regarding the `lowmem` solution from GOG, here is the report from one of our users:

`Did some additional testing. Found a "solution" in the bowels of the GoG forums. Starting the game with -LOWMEM lets you pass the scene and the interior loads, and Baccata can have his tirade. Nevermind, GOG version crashes halfway into Baccata's lines with the same error. 
I thought it might be the resolution, as lowmem defaults back to 320x200, but just lowering that option still crashes, so there is something with the system memory, but I couldn't find anything else sadly.`

So even (or unsurprisingly?) the GOG version suffers this problem and there is no real solution.

Ive tried Staging with memsize=8 all the way through 8mb steps up to 64mb with no change to the error or even the numbers being reported in it if that helps.

I am *really* looking forward to @boohyaka results

**Comment by Python-Exoproject** [comment]
For anyone testing this in dosbox (or maybe useful to boohyaka) here are the only two glide2.ovl files that let the game start using ECE or Staging in 3dfx mode: [Voodoo.zip](https://github.com/dosbox-staging/dosbox-staging/files/14544928/Voodoo.zip)

**Comment by boohyaka** [comment]
Exact same bug using the provided safefile on a P233 MMX with Voodoo1 @Python-Exoproject!

**Comment by johnnovak** [comment]
> Exact same bug using the provided savefile on a P233 MMX with Voodoo1 @Python-Exoproject!

Excellent @boohyaka, thank you very much for confirming this is indeed a game bug. Closing.

**Comment by Python-Exoproject** [comment]
For those interested a fix has been posted for the game at https://github.com/kjliew/qemu-3dfx/issues/20#issuecomment-1984851882 which ive now tested and confirmed works. 

Heres the fixed exe: [LOLG.zip](https://github.com/dosbox-staging/dosbox-staging/files/14557632/LOLG.zip)

**Comment by boohyaka** [comment]
I confirm it works on real hardware too! Good stuff.

**Comment by kcgen** [comment]
Awesome progress guys :-)

Added a couple other crash fixes (as alternatives) and a CDROM full-install QoL improvement:

https://github.com/exoscoriae/eXoDOS/issues/5055#issuecomment-1989278843

**Comment by boohyaka** [comment]
Hey, awesome to read from you mate :) and good job too! the full cdrom install is great!

**Comment by johnnovak** [comment]
Nice one @kcgen and glad to see you commenting again 😄

**Comment by johnnovak** [comment]
> Thank you for your hard work on Lands of Lore: GoD. In case anyone experience issues when trying to run the game on macOS, feel free to follow my guide on [how to run the game on newest macOS Sequoia with 3dfx](https://mattkomarnicki.com/articles/running-lands-of-lore-guardians-of-destiny-on-macos-sequoia-with-3dfx).

That's a nice guide @mattkomarnicki. Do note you can use per-game configs on macOS as well (plus our start icon), and you could use drive automounts as explained in our Getting Started guide.
